compiler/src/iree/compiler/Dialect/Util/Transforms/TraceFrameMark.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/base/tracing.h"
+#include "iree/compiler/Dialect/Util/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Util {
+
+namespace {
+
+class TraceFrameMarkBeginPass
+    : public TraceFrameMarkBeginBase<TraceFrameMarkBeginPass> {
+ public:
+  TraceFrameMarkBeginPass() = default;
+  TraceFrameMarkBeginPass(TraceFrameName name) { this->name = name; }
```

Comment:
we can avoid the enum and take a StringRef instead - the literal should then be captured when the pass is created instead of needing to specialize it all here - that would make the passes generic

Reply:
Ah, I was assuming that I needed a compile-time-constant literal (without any indirection through C++ classes). If I can just pass StringRef that would be great. I'll give it a try!

Reply:
Tested a bit, and the critical detail is that the string should be null-terminated and the same pointer should be used for Begin and End.

`llvm::StringRef` works if you're careful. I could also add a wrapper that ensures they use the same pointer

```c++
// These patterns should both work, for example:
createTraceFrameMarkBeginPass("Foo");
createTraceFrameMarkEndPass("Foo");

const char phaseName[] = "Foo\0";
createTraceFrameMarkBeginPass(phaseName);
createTraceFrameMarkEndPass(phaseName);

// This will *not* work:
std::string phaseNameBegin = "Foo";
std::string phaseNameEnd = "Foo";
createTraceFrameMarkBeginPass(phaseNameBegin);
createTraceFrameMarkEndPass(phaseNameEnd);
```

A wrapper could look like
```c++
TraceFrame traceFrame("Foo");
passManager.addPass(traceFrame.createMarkBeginPass());
passManager.addPass(traceFrame.createMarkEndPass());
```